File: src/main/java/xyz/phanta/cbtweaker/hatch/HatchManager.java
package xyz.phanta.cbtweaker.hatch;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import io.github.phantamanta44.libnine.util.helper.JsonUtils9;
import net.minecraft.util.ResourceLocation;
import xyz.phanta.cbtweaker.CbtMod;
import xyz.phanta.cbtweaker.ItemBlockRegistrar;
import xyz.phanta.cbtweaker.buffer.BufferType;
import xyz.phanta.cbtweaker.buffer.BufferTypeRegistry;
import xyz.phanta.cbtweaker.common.BlockMaterial;
import xyz.phanta.cbtweaker.gui.CbtTextureResources;
import xyz.phanta.cbtweaker.gui.inventory.UiLayout;
import xyz.phanta.cbtweaker.util.ConfigException;
import xyz.phanta.cbtweaker.util.helper.DataLoadUtils;

import javax.annotation.Nullable;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class HatchManager {

    private static final UiLayout DEFAULT_UI_LAYOUT = new UiLayout(CbtTextureResources.GUI_GENERIC_SMALL,
            UiLayout.DEFAULT_MACHINE_REGION, true, UiLayout.DEFAULT_PLAYER_INV_POS, true);

    private final Path specDir;
    private final Map<String, HatchType<?>> hatchTypeTable = new HashMap<>();
    private final ItemBlockRegistrar registrar;

    public HatchManager(Path specDir, ItemBlockRegistrar registrar) {
        this.specDir = specDir;
        this.registrar = registrar;
    }

    public void loadAll() {
        CbtMod.LOGGER.info("Loading hatch types...");
        try {
            Files.list(specDir).forEach(specFile -> {
                try {
                    if (!Files.isRegularFile(specFile)) {
                        return;
                    }
                    String specFileName = specFile.getFileName().toString();
                    if (!specFileName.endsWith(".json")) {
                        CbtMod.LOGGER.warn("Ignoring non-JSON file in hatch specification directory: {}", specFileName);
                        return;
                    }
                    String hatchId = specFileName.substring(0, specFileName.length() - 5);
                    if (hatchId.isEmpty()) {
                        throw new ConfigException("Bad hatch ID: " + hatchId);
                    }

                    JsonObject specDto = DataLoadUtils.readJsonFile(specFile).getAsJsonObject();
                    BlockMaterial blockMat = specDto.has("material")
                            ? BlockMaterial.fromString(specDto.get("material").getAsString()) : BlockMaterial.METAL;

                    ResourceLocation bufTypeId = new ResourceLocation(specDto.get("type").getAsString());
                    BufferTypeRegistry.Entry<?, ?, ?, ?> bufTypeEntry = CbtMod.PROXY.getBufferTypes().lookUp(bufTypeId);
                    if (bufTypeEntry == null) {
                        throw new ConfigException("Unknown buffer type: " + bufTypeId);
                    }

                    JsonObject archetypeDto = specDto.has("archetype")
                            ? specDto.getAsJsonObject("archetype") : new JsonObject();
                    JsonArray tiersDto = specDto.getAsJsonArray("tiers");
                    if (tiersDto == null || tiersDto.size() == 0) {
                        throw new ConfigException("At least one tier must be defined!");
                    } else if (JsonUtils9.stream(tiersDto).anyMatch(t -> !t.isJsonObject())) {
                        throw new ConfigException("Tier config must be a JSON object!");
                    }

                    HatchType<?> hatchType = constructHatchType(hatchId, blockMat, bufTypeEntry.getBufferType(),
                            JsonUtils9.stream(tiersDto)
                                    .map(t -> JsonUtils9.mergeCloning(archetypeDto, t.getAsJsonObject())));
                    hatchTypeTable.put(hatchId, hatchType);
                    registrar.addBlockRegistration(
                            "hatch_" + hatchId, HatchBlock.construct(hatchType), HatchBlockItem::new);
                    CbtMod.LOGGER.debug("Loaded hatch: {}", hatchId);
                } catch (Exception e) {
                    throw new IllegalStateException("Failed to load hatch specification: " + specFile, e);
                }
            });
        } catch (Exception e) {
            throw new IllegalStateException("Failed to load hatch types!", e);
        }
        CbtMod.LOGGER.info("Finished loading hatch specifications.");
    }

    private static <B> HatchType<B> constructHatchType(String id, BlockMaterial blockMat, BufferType<B, ?, ?, ?> bufType,
                                                       Stream<JsonObject> tierConfigs) {
        return new HatchType<>(id, blockMat, bufType, tierConfigs
                .map(tierConfig -> new HatchType.TierData<>(
                        bufType.loadBufferFactory(
                                tierConfig.has("buffer") ? tierConfig.getAsJsonObject("buffer") : new JsonObject()),
                        tierConfig.has("ui")
                                ? CbtMod.PROXY.getTemplates().getUiTemplates().resolve(tierConfig.get("ui"))
                                : DEFAULT_UI_LAYOUT))
                .collect(Collectors.toList()));
    }

    @Nullable
    public HatchType<?> lookUp(String id) {
        return hatchTypeTable.get(id);
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/hatch/HatchManager.kt
package st.evening.mc.cbtweaker.hatch

import net.minecraft.util.ResourceLocation
import st.evening.mc.cbtweaker.CbTweaker
import st.evening.mc.cbtweaker.buffer.BufferType
import st.evening.mc.cbtweaker.common.BlockConfig
import st.evening.mc.cbtweaker.gui.inventory.WindowConfig
import st.evening.mc.cbtweaker.util.DataGenHelper
import st.evening.mc.cbtweaker.util.forEachFile
import st.evening.mc.prelude.api.data.ser.SerializationException
import st.evening.mc.prelude.api.data.tjson.JsonPath
import st.evening.mc.prelude.api.data.tjson.TJson
import st.evening.mc.prelude.api.data.tjson.TypedJsonParser
import st.evening.mc.prelude.api.data.tjson.expectObject
import st.evening.mc.prelude.api.data.tjson.forEachObject
import st.evening.mc.prelude.api.data.tjson.useAny
import st.evening.mc.prelude.api.data.tjson.useArrayValue
import st.evening.mc.prelude.api.data.tjson.useObject
import st.evening.mc.prelude.api.data.tjson.useStringValue
import st.evening.mc.prelude.api.registration.ModRegistrar
import st.evening.mc.prelude.api.util.data.TypedJsonHelper
import java.nio.file.Files
import java.nio.file.Path
import kotlin.io.path.readText

class HatchManager(private val specDir: Path, private val reg: ModRegistrar) : Iterable<HatchType<*>> {
    private val hatchTypeTable: MutableMap<String, HatchType<*>> = mutableMapOf()

    operator fun get(id: String): HatchType<*>? = hatchTypeTable[id]

    override fun iterator(): Iterator<HatchType<*>> = hatchTypeTable.values.iterator()

    fun loadAll() {
        CbTweaker.logger.info("Loading hatch types...")
        try {
            specDir.forEachFile { specFile ->
                try {
                    if (!Files.isRegularFile(specFile)) return@forEachFile
                    val specFileName = specFile.fileName.toString()
                    if (!specFileName.endsWith(".tjson")) {
                        CbTweaker.logger.warn(
                            "Ignoring non-TJSON file in hatch specification directory: {}",
                            specFileName
                        )
                        return@forEachFile
                    }
                    val hatchId = specFileName.dropLast(6)
                    if (hatchId.isBlank()) {
                        throw SerializationException("Empty hatch ID!")
                    }

                    val specDto = TypedJsonParser.parseObject(specFile.readText())
                    JsonPath.atRoot {
                        val bufTypeEntry = specDto.useStringValue("type") {
                            CbTweaker.defns.bufferTypes[ResourceLocation(it)]
                                ?: throw SerializationException.withPath("Unknown buffer type: $it")
                        }
                        hatchTypeTable[hatchId] = loadHatchType(hatchId, bufTypeEntry.bufferType, specDto)
                    }
                    CbTweaker.logger.debug("Loaded hatch: {}", hatchId)
                } catch (e: Exception) {
                    throw IllegalStateException("Failed to load hatch specification: $specFile", e)
                }
            }
        } catch (e: Exception) {
            throw IllegalStateException("Failed to load hatch types!", e)
        }
        CbTweaker.logger.info("Loaded {} hatch types", hatchTypeTable.size)
    }

    context(_: JsonPath)
    private fun <B> loadHatchType(
        id: String,
        bufType: BufferType<B, *, *, *>,
        specDto: TJson.Object
    ): HatchType<B> {
        val blockConfig = specDto.useObject("block") { BlockConfig.load(it) } ?: BlockConfig.DEFAULT
        val archetypeDto = specDto.expectObject("archetype")
        val tiers = specDto.useArrayValue("tiers") { tiersDto ->
            buildList {
                tiersDto.forEachObject { tierDtoRaw ->
                    val tierDto = merge(archetypeDto, tierDtoRaw)
                    add(
                        HatchType.TierData(
                            bufType.loadBufferFactory(tierDto.expectObject("buffer") ?: TJson.Object()),
                            tierDto.useAny("window") {
                                CbTweaker.defns.templates.windowTemplates.resolve(it)
                            } ?: WindowConfig.GENERIC_SMALL
                        )
                    )
                }
                if (isEmpty()) throw SerializationException.withPath("At least one tier must be defined!")
            }
        }
        return HatchType(reg, id, blockConfig, bufType, tiers)
    }

    private fun merge(archetype: TJson.Object?, instance: TJson.Object): TJson.Object =
        archetype?.let { TypedJsonHelper.merge(it, instance) } ?: instance

    fun dataGenBlockModels(resourceDir: Path) {
        if (hatchTypeTable.isEmpty()) return
        val cbtDir = resourceDir.resolve("cbtweaker")
        val blockStateDir = Files.createDirectories(cbtDir.resolve("blockstates"))
        val itemModelDir = cbtDir.resolve("models/item")
        hatchTypeTable.forEach { (id, hatchType) ->
            val bsFile = blockStateDir.resolve("hatch_$id.json")
            if (!Files.exists(bsFile)) {
                CbTweaker.logger.info("Generating hatch block state mapping: {}", bsFile.fileName)
                DataGenHelper.writeToFile(bsFile) {
                    "forge_marker" number 1
                    "variants" obj {
                        "tier" obj {
                            for (i in 0..<hatchType.tierCount) {
                                i.toString() obj {
                                    "model" string "cbtweaker:hatch_${i.coerceAtMost(3)}"
                                }
                            }
                        }
                    }
                }
            }
            val hatchItemModelDir = Files.createDirectories(itemModelDir.resolve("hatch_$id"))
            for (i in 0..<hatchType.tierCount) {
                val tierFile = hatchItemModelDir.resolve("hatch_${id}_$i.json")
                if (!Files.exists(tierFile)) {
                    CbTweaker.logger.info("Generating hatch item model: {}", tierFile.fileName)
                    DataGenHelper.writeToFile(tierFile) {
                        "parent" string "cbtweaker:block/hatch_${i.coerceAtMost(3)}"
                    }
                }
            }
        }
    }
}
